File: src/main/java/com/tabnine/inline/render/experimental/InlineElementRenderer.java
package com.tabnine.inline.render.experimental;

import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.editor.EditorCustomElementRenderer;
import com.intellij.openapi.editor.Inlay;
import com.intellij.openapi.editor.markup.TextAttributes;
import com.intellij.ui.JBColor;
import org.jetbrains.annotations.NotNull;
import java.awt.*;
import static com.tabnine.inline.render.GraphicsUtils.getFont;
import static com.tabnine.inline.render.GraphicsUtils.getNiceContrastColor;

public class InlineElementRenderer implements EditorCustomElementRenderer {
    private final Editor editor;
    private final String suffix;
    private final boolean deprecated;
    private Integer renderingXAnchor;
    private Color color;

    public InlineElementRenderer(Editor editor, String suffix, boolean deprecated) {
        this.editor = editor;
        this.suffix = suffix;
        this.deprecated = deprecated;
    }

    @Override
    public int calcWidthInPixels(@NotNull Inlay inlay) {
        return this.editor.getContentComponent()
                .getFontMetrics(getFont(this.editor, this.deprecated)).stringWidth(this.suffix);
    }

    @Override
    public void paint(
            @NotNull Inlay inlay,
            @NotNull Graphics g,
            @NotNull Rectangle targetRegion,
            @NotNull TextAttributes textAttributes) {
        renderingXAnchor = renderingXAnchor == null ? targetRegion.x : renderingXAnchor;
        color = color == null ? getNiceContrastColor() : color;
        g.setColor(color);
        g.setFont(getFont(this.editor, this.deprecated));
        g.drawString(suffix, renderingXAnchor, targetRegion.y + editor.getAscent());
    }
}

File: src/main/java/com/tabnine/inline/render/experimental/InlineElementRenderer.kt
package com.tabnine.inline.render.experimental

import com.intellij.openapi.editor.Editor
import com.intellij.openapi.editor.EditorCustomElementRenderer
import com.intellij.openapi.editor.Inlay
import com.intellij.openapi.editor.markup.TextAttributes
import com.tabnine.inline.render.GraphicsUtils
import java.awt.Color
import java.awt.Graphics
import java.awt.Rectangle

class InlineElementRenderer(private val editor: Editor, private val suffix: String, private val deprecated: Boolean) :
    EditorCustomElementRenderer {
    private var renderingXAnchor: Int? = null
    private var color: Color? = null
    override fun calcWidthInPixels(inlay: Inlay<*>): Int {
        return editor.contentComponent
            .getFontMetrics(GraphicsUtils.getFont(editor, deprecated)).stringWidth(suffix)
    }

    override fun paint(
        inlay: Inlay<*>,
        g: Graphics,
        targetRegion: Rectangle,
        textAttributes: TextAttributes
    ) {
        renderingXAnchor = renderingXAnchor ?: targetRegion.x
        color = color ?: GraphicsUtils.niceContrastColor
        g.color = color
        g.font = GraphicsUtils.getFont(editor, deprecated)
        g.drawString(suffix, renderingXAnchor!!, targetRegion.y + editor.ascent)
    }
}
